src/udev/scsi_id/scsi_serial.c

```diff
@@ -85,8 +89,111 @@ static const char hex_str[]="0123456789abcdef";
+                                               int len);
+static void process_std_inquiry_data(struct scsi_id_device *dev_scsi, unsigned char *buf);
+
+static int sgio_do_scsi_page80_inquiry(struct scsi_id_device *dev_scsi, int fd,
+                                       char *serial, char *serial_short, int max_len);
+static void scsi_get_dev_name(const char *filePath, char *dev_name);
+static void scsi_get_device_vpd_page_path(const char *device_path, char *device_vpd_page_path, unsigned int page);
+
+static void scsi_get_dev_name(const char *file_path, char *dev_name) {
+        _cleanup_free_ char *file_name = NULL;
+        int r;
+
+        r = path_extract_filename(file_path, &file_name);
+
+        if (r < 0 || r == O_DIRECTORY) {
+                printf("Invalid Path Specified\n");
+                exit(0);
+        }
+        strcpy(dev_name, file_name);
+}
+
+static void scsi_get_device_vpd_page_path(const char *device_path, char *device_vpd_page_path, unsigned int page) {
+        char dev_name[MAX_ATTR_LEN];
+        memset(dev_name, 0, MAX_ATTR_LEN);
+
+        scsi_get_dev_name(device_path, dev_name);
+        strcpy(device_vpd_page_path, PAGE_PREFIX);
+        strcat(device_vpd_page_path, dev_name);
+
+        switch(page) {
+                case PAGE_UNSPECIFIED:
+                       strcat(device_vpd_page_path, PAGE0x00_SUFFIX);
+                       break;
+                case PAGE_80:
+                        strcat(device_vpd_page_path, PAGE0x80_SUFFIX);
+                        break;
+                case PAGE_83:
+                        strcat(device_vpd_page_path, PAGE0x83_SUFFIX);
+                        break;
+                default:
+                        printf("Invalid page code\n");
+                        exit(0);
+        }
+}
+
+
+int scsi_get_serial_open_file(const char *devname, int flags, int page_code) {
+        int cnt;
+        int fd = -1;
+        char file_name[MAX_PATH_LEN];
+        char device_vpd_page_path[MAX_PATH_LEN];
+
+        memset(file_name, 0, MAX_PATH_LEN);
+        memset(device_vpd_page_path, 0, MAX_PATH_LEN);
+
+        switch(scsi_interface) {
+                case SCSI_ID_INTERFACE_SGIO:
+                        strcpy(file_name, devname);
+                        break;
+                case SCSI_ID_INTERFACE_SYSFS:
+                case SCSI_ID_INTERFACE_UNKNOWN:
+                        scsi_get_device_vpd_page_path(devname, device_vpd_page_path, page_code);
+                        strcpy(file_name, device_vpd_page_path);
+                        break;
+                default:
+                       assert_not_reached();
+        }
+
+        for (cnt = 20; cnt > 0; cnt--) {
+                fd = RET_NERRNO(open(file_name, O_RDONLY | O_NONBLOCK | O_CLOEXEC));
+                if (fd >= 0 || fd != -EBUSY)
+                        return fd;
+
+                (void)usleep((200 * USEC_PER_SEC) + (random_u64_range(100 * USEC_PER_SEC)));
```

Comment:
Please use `usleep_safe()`, and insert a space between `(void)` and the function name.

Reply:
Done